File: egghead.io/data-structures-and-algorithms-in-js/lesson-08-write-a-dfs-search-algorithm-for-graphs-in-js/search_graph.py
import os
import sys
sys.path.append(
  '/'.join(os.getcwd().split('/')[:-1] + [
    'lesson-06-js-graph-data-structure'
  ])
)
sys.path.append(
  '/'.join(os.getcwd().split('/')[:-1] + [
    'lesson-02-queue-data-structure-in-js'
  ])
)

from graph import Graph
from custom_queue import  CustomQueue


class SearchGraph(Graph):
  def breadth_first_search(self, key, func):
    head = self.get_node(key)
    hist = { node.key: False for node in self.nodes }
    custom_queue = CustomQueue()
    custom_queue.enqueue(head)

    while not custom_queue.is_empty():
      curr = custom_queue.dequeue()

      if not hist[curr.key]:
        func(curr)
        hist[curr.key] = True
      
      for neighbor in curr.neighbors:
        if not hist[neighbor.key]:
          custom_queue.enqueue(neighbor)
  
  def depth_first_search(self, key, func):
    head = self.get_node(key)
    hist = { node.key: False for node in self.nodes }

    def traverse(node):
      if hist[node.key]: return

      func(node)
      hist[node.key] = True

      for neighbor in node.neighbors:
        traverse(neighbor)
    
    traverse(head)    

File: egghead.io/data-structures-and-algorithms-in-js/lesson-08-write-a-dfs-search-algorithm-for-graphs-in-js/search-graph.js
const Graph = require('../lesson-06-js-graph-data-structure/graph');
const CustomQueue = require('../lesson-02-queue-data-structure-in-js/custom-queue');


class SearchGraph extends Graph {
  breadthFirstSearch(key, func) {
    const head = this.getNode(key);
    const hist = this.nodes.reduce((acc, node) => {
      acc[node.key] = false;
      return acc;
    }, {});
    const customQueue = new CustomQueue();
    customQueue.enqueue(head);

    while (!customQueue.isEmpty()) {
      const curr = customQueue.dequeue();

      if (!hist[curr.key]) {
        func(curr);
        hist[curr.key] = true;
      }

      curr.neighbors.map(neighbor => {
        if (!hist[neighbor.key]) {
          customQueue.enqueue(neighbor);
        }
      })
    }
  }

  depthFirstSearch(key, func) {
    const head = this.getNode(key);
    const hist = this.nodes.reduce((acc, node) => {
      acc[node.key] = false;
      return acc;
    }, {});

    function traverse(node) {
      if (hist[node.key]) {
        return
      }

      func(node);
      hist[node.key] = true;

      node.neighbors.map(neighbor => traverse(neighbor));
    }

    traverse(head);
  }
}

module.exports = SearchGraph;
